src/Controller/ContentTypeController.php

```diff
@@ -0,0 +1,107 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Webmozart\Assert\Assert;
+
+final readonly class ContentTypeController
+{
+    /**
+     *  \p{L} → Matches any letter (Latin, Kanji, Hiragana, Katakana, etc.)
+     *  \p{N} → Matches any number (so numbers remain valid in slugs)
+     *  (?:-[\p{L}\p{N}]+)* → Allows hyphenated words (e.g., hello-world)
+     *  (?:\/[\p{L}\p{N}]+(?:-[\p{L}\p{N}]+)*)*\/? → Allows slashes for hierarchical paths
+     *  Trailing slash (\/?) → Optional to allow both /slug and /slug/.
+     */
+    public const string PATH_PATTERN = '([\p{L}\p{N}]+(?:-[\p{L}\p{N}]+)*(?:\/[\p{L}\p{N}]+(?:-[\p{L}\p{N}]+)*)*\/?)$';
+    private Version $version;
+
+    public function __construct(
+        private ContentTypeControllerRegistry $registry,
+        private ContainerInterface $container,
+        private LoggerInterface $logger,
+        private RequestHandlerInterface $requestHandler,
+        string $version,
+    ) {
+        $this->version = Version::from($version);
+    }
+
+    public function __invoke(Request $request, string $slug): Response
+    {
+        $response = $this->requestHandler->handle($request, $slug);
+
+        if ($response instanceof Response) {
+            return $response;
+        }
+
+        $story = $response->story;
+        Assert::keyExists($story, 'published_at');
+        Assert::keyExists($story, 'content');
+        Assert::keyExists($story['content'], 'component');
+
+        try {
+            $definition = $this->registry->bySlug(\sprintf('/%s', $slug));
+            $this->logger->debug('Found content type controller by slug.', [
+                'slug' => $slug,
+                'controller' => $definition->className,
+            ]);
+        } catch (ContentTypeControllerNotFoundException) {
+            $type = TrimmedNonEmptyString::fromString($story['content']['component'])->toString();
+            $definition = $this->registry->byType($type);
+            $this->logger->debug('Found content type controller by component.', [
+                'type' => $type,
+                'controller' => $definition->className,
+            ]);
+        }
+
+        /** @var callable(Request, ContentTypeInterface): Response $controller */
+        $controller = $this->container->get($definition->className);
+
+        /** @var ContentTypeInterface $contentType */
+        $contentType = new $definition->dto($story);
+
+        $this->logger->debug('Calling content type controller.', [
+            'controller' => $definition->className,
+            'content_type' => $contentType,
+        ]);
+
+        $response = $controller($request, $contentType);
+
+        $this->logger->debug('Content type controller returned response.', [
+            'controller' => $definition->className,
+            'response' => $response,
+        ]);
+
+        if ($this->version->equals(Version::Published)) {
```

Comment:
For now it is ok to set proper cache values, but one would like to change that in the future. 

Out of scope here, but we could make this configurable for the user

Reply:
I think that fits the scope :) and i think a great idea!

Reply:
Done!